data/blog/mobile-app-launch-profiling.md

```diff
@@ -0,0 +1,46 @@
+---
+title: 'Mobile App Launch Profiling'
+date: '2024-04-17'
+tags: ['mobile', 'ios', 'android', 'profiling']
+draft: false
+summary: "See what's happening in your app before your first line of code can even run."
+images: ['/images/mobile-app-launch-profiling/launch-profile.png']
+layout: PostLayout
+canonicalUrl: mobile-app-launch-profiling
+authors: ['andrewmcknight']
+---
+
+If you use Sentry’s Performance monitoring on Cocoa and Android, you also have the ability to see [profiles](https://docs.sentry.io/product/profiling/) of how your app runs in the wild on real end user devices. Any time an automatic or manual transaction is recorded, that profiling data is attached. But are there times where you can’t record a transaction? 🤔
+
+# App launch performance
+
+<aside>
+💡 This post will focus on our Cocoa app launch profiling feature. You can read more about the Android counterpart by heading over to that area of our docs: https://docs.sentry.io/platforms/android/profiling/#app-start-profiling
+</aside>
+
+Apple places special focus on good [performance of an app’s launch](https://developer.apple.com/documentation/xcode/reducing-your-app-s-launch-time). They even built an entire process to enforce good performance: the [watchdog](https://developer.apple.com/documentation/xcode/addressing-watchdog-terminations). There are two ways you can monitor launch performance today:
+
+1. Instruments. You can profile your app while launching it on any devices you have available.
+2. MetricKit. You can see a histogram of app launch times in Xcode’s Organizer. This data is inexact, because you can’t correlate any specific context to problematic launches, and the way launch time is measured is out of your control:
+
+> The launch-time metric measures the time from the user tapping the app icon on their Home screen to the app drawing its first frame to the screen. If your app still has to run code after it has drawn its first frame, but before the user can begin using the app, that time doesn’t contribute to the launch-time metric. **Extra startup activities still contribute to the user’s perception of the app’s responsiveness**.
+
+So, if you or one of your dependencies does some work at launch time, you don’t necessarily see that in the Xcode Organizer.
```

Comment:
Is it so that it only measures up to showing the spinner? 
I imagine this is common enough concept that could help folks understand this: 

"Basically this is telling you how quickly you're rendering your loading spinner"

Reply:
A loading spinner is a specific case, but not everybody necessarily has one. I'm not sure this is necessary to add in.

Reply:
Even if their app doesn't have a spinner, the concept is very known since apps everyone uses have them. I suggested only to help others undersand (it took me a second to understand what this meant.

Reply:
So we're actually talking about work being done even before any UI is visible, including something like a loading spinner. I will expand on this a bit.

Reply:
Yeah you're right, with the more recent pattern being an "empty state" that has some animation.

Reply:
see https://github.com/getsentry/sentry.engineering/pull/124/files#r1779271955